frontend/lib/src/components/shared/Toolbar/SharedToolbar.tsx

```diff
@@ -0,0 +1,84 @@
+/**
+ * Copyright (c) Streamlit Inc. (2018-2022) Snowflake Inc. (2022-2024)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import React, {
+  createContext,
+  ReactNode,
+  useContext,
+  useEffect,
+  useState,
+} from "react"
+
+import Toolbar, {
+  ToolbarAction as BaseToolbarAction,
+  ToolbarActionProps,
+} from "./Toolbar"
+import { StyledContentWrapper } from "./styled-components"
+
+const ToolbarContext = createContext<{
+  actions: ToolbarActionProps[]
+  setActions: React.Dispatch<React.SetStateAction<ToolbarActionProps[]>>
+} | null>(null)
+
+export const ToolbarRenderer = ({
+  children,
+}: {
+  children: ReactNode
+}): React.ReactElement => {
+  const [actions, setActions] = useState<ToolbarActionProps[]>([])
+
+  return (
+    <ToolbarContext.Provider value={{ actions, setActions }}>
+      <StyledContentWrapper>
+        <Toolbar target={StyledContentWrapper}>
```

Comment:
Not sure if the current state of abstraction will work for most of our toolbar uses. In many cases, we might need more fine-grained handling of the toolbar behavior (see the locked & fullscreen-related properties):
<img width="394" alt="image" src="https://github.com/user-attachments/assets/58ffa595-c75d-434d-b142-fccdf23468b3">

Also, it might not always be the case that we want to trigger the toolbar on the full element div -> e.g., in case of audio input, we currently only trigger it if the wave container is hovered, not if the label is hovered. 

Based on the current plans, the toolbar's main usage will be for components that also provide a fullscreen mode. If we add such an abstraction, it might be important to cover the fullscreen case with the abstraction as well.

Reply:
In terms of all of these other properties, they could be configured via the context of the widget. It would take some more refactoring, but could be done! 

In terms of the full element div, that's a really good call out and something we should support! I think the context could also expose an option that allows the portal location (note: I am referencing my portal suggestion in my review below in order to make this work) to be overridden to be any arbitrary `ref`

Reply:
> In terms of all of these other properties, they could be configured via the context of the widget.
What would that look like? Would it be a useContext that has some `setLocked` `setDisableFullscreenMode` and such? I'd hope we could have a more declarative approach... perhaps a useToolbarOptions({locked: ...}) that handles all that for us?

Reply:
I made some updates on my own branch to show how this would work (yes, basically `useToolbarOptions` like you said): https://github.com/streamlit/streamlit/compare/develop...bnisco/shared-toolbar-and-widget-label

This concept continues to push shared, common components above the widgets, allowing individual widgets to focus on their business-specific logic. Ideally, each widget should not need to explicitly render shared concepts (such as a `Toolbar` or a `FullscreenWrapper`). Instead, they should call a hook or render a business-logic specific component that integrates with and augments the existing concepts when needed. This provides an incremental on-ramp and brings the amount of duplicated code down substantially from widget to widget.

Reply:
I pushed another few updates to my branch. This latest revision also includes hoisting all of the Fullscreen logic out of the widget and into a component that is rendered above each widget. To show how this works, I migrated the DeckGL component to use it. One other side-effect benefit to this approach is that there is a lot less prop drilling, since things can just read from the widget-scoped context and get the correct information as needed.

This was great since I went down the rabbit hole of fullscreen and found a few things that can be improved to help with performance (specifically, we are re-measuring window dimensions in many components unnecessarily right now), even if we don't end up using the general strategy proposed there.